user_linux_cgo.go

```diff
@@ -0,0 +1,216 @@
+// +build cgo
+// +build linux
+
+package buildah
+
+// #include <sys/types.h>
+// #include <grp.h>
+// #include <pwd.h>
+// #include <stdlib.h>
+// #include <stdio.h>
+// #include <string.h>
+// typedef FILE * pFILE;
+import "C"
+
+import (
+	"fmt"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strings"
+	"sync"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+func fopenContainerFileWork(rootdir, filename string, linkCount int) (string, C.pFILE, error) {
+	// If we've recursed too far, return an error.
+	if linkCount >= 64 {
+		return "", nil, fmt.Errorf("too many links encountered looking up %q", filename)
+	}
+
+	// Do a bit of lexical cleanup on the filename.
+	filename = filepath.Clean(filename)
+
+	// Start the lookup at the chroot directory.
+	rootfd, err := unix.Open(rootdir, unix.O_DIRECTORY|unix.O_PATH, 0)
+	if err != nil {
+		return "", nil, err
+	}
+	defer unix.Close(rootfd)
+
+	// Start resolving the pathname.
+	components := strings.Split(filepath.Clean(filename), string(os.PathSeparator))
```

Comment:
Why are you cleaning filename a second time?

Reply:
Yeah, that's a mistake.